x/exchange/keeper/grpc_query.go

```diff
@@ -545,3 +545,172 @@
+				"source", source.String(), "value", fmt.Sprintf("%v", value),
+				"keyPrefix", fmt.Sprintf("%v", keyPrefix), "keySuffix", fmt.Sprintf("%v", keySuffix))
+			return nil
+		}
+		if payment == nil {
+			k.logEndpointError(ctx, "GetPaymentsWithSource", "Empty payment entry.",
+				"source", source.String(), "value", fmt.Sprintf("%v", value),
+				"keyPrefix", fmt.Sprintf("%v", keyPrefix), "keySuffix", fmt.Sprintf("%v", keySuffix))
+			return nil
+		}
+		resp.Payments = append(resp.Payments, payment)
+		return nil
+	})
+
+	if pageErr != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "error iterating payments with source %s: %v", req.Source, pageErr)
+	}
+
+	return resp, nil
+}
+
+// GetPaymentsWithTarget gets all payments with a specific target account.
+func (k QueryServer) GetPaymentsWithTarget(goCtx context.Context, req *exchange.QueryGetPaymentsWithTargetRequest) (*exchange.QueryGetPaymentsWithTargetResponse, error) {
+	if req == nil || len(req.Target) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+	target, err := sdk.AccAddressFromBech32(req.Target)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid target %q: %v", req.Target, err)
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	keyPrefix := GetIndexKeyPrefixTargetToPayments(target)
+	store := k.getStore(ctx)
+	preStore := prefix.NewStore(store, keyPrefix)
+
+	resp := &exchange.QueryGetPaymentsWithTargetResponse{}
+	var pageErr error
+	resp.Pagination, pageErr = query.Paginate(preStore, req.Pagination, func(keySuffix, _ []byte) error {
+		// Only add it to the result if we can read it. This might result in fewer results than the limit,
+		// but at least one bad entry won't block others by causing the whole thing to return an error.
+		source, externalID, pErr := ParseIndexKeySuffixTargetToPayment(keySuffix)
+		if pErr != nil {
+			k.logEndpointError(ctx, "GetPaymentsWithTarget", "Error reading target to payment index entry.",
+				"error", pErr, "target", target.String(),
+				"keyPrefix", fmt.Sprintf("%v", keyPrefix), "keySuffix", fmt.Sprintf("%v", keySuffix))
+			return nil
+		}
+
+		payment, pErr := k.getPaymentFromStore(store, source, externalID)
+		if pErr != nil {
+			k.logEndpointError(ctx, "GetPaymentsWithTarget", "Error reading payment from store.", "error", pErr,
+				"target", target.String(), "source", source.String(), "externalID", externalID,
+			)
+			return nil
+		}
+		if payment == nil {
+			k.logEndpointError(ctx, "GetPaymentsWithTarget", "No payment found from target to payment index entry.",
+				"target", target.String(), "source", source.String(), "externalID", externalID)
+			return nil
+		}
+
+		resp.Payments = append(resp.Payments, payment)
+		return nil
+	})
+
+	if pageErr != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "error iterating payments with target %s: %v", req.Target, pageErr)
+	}
+
+	return resp, nil
+}
```

Comment:
Everything provided to the chain is public information. The only way sensitive information would end up in the logs is if a user provides that sensitive information. We do what we can to make sure users know that everything on chain is public and they shouldn't give it anything they don't want public. And if they do provide sensitive information to the chain, they have bigger problems than the data being in the logs.

The `logEndpointError` function is about the best that can be done to abstract common logic. Additionally, I wouldn't be surprised if these log messages are ever encountered. So, for the purposes of optimization, I don't want to add any code that gets executed outside these cases that is only needed for the logging. I.e. I don't want to put a logger in a variable at the start of the endpoint because I don't expect it to ever actually be used, and I don't want all that extra processing done all the time.